src/goto-programs/validate_goto_model.cpp

```diff
@@ -0,0 +1,237 @@
+  {
+    for(auto &instr : fun.second.body.instructions)
+    {
+      if(instr.is_function_call())
+      {
+        const code_function_callt &function_call =
+          to_code_function_call(instr.code);
+        DATA_CHECK(
+          function_call.function().id() == ID_symbol,
+          "no calls via fn pointer should be present");
+      }
+    }
+  }
+}
+
+void validate_goto_modelt::check_returns_removed()
+{
+  for(const auto &fun : function_map)
+  {
+    const goto_functiont &goto_function = fun.second;
+    DATA_CHECK(
+      goto_function.type.return_type().id() == ID_empty,
+      "functions must have empty return type");
+
+    for(const auto &instr : goto_function.body.instructions)
+    {
+      DATA_CHECK(
+        !instr.is_return(), "no return instructions should be present");
+
+      if(instr.is_function_call())
+      {
+        const auto &function_call = to_code_function_call(instr.code);
+        DATA_CHECK(
+          function_call.lhs().is_nil(), "function call return should be nil");
+
+        // consider: if this is redundant?
+        // I could not trigger this with unit tests without failing
+        // the first test "goto_function.type.return_type().id() == ID_empty"
+        // (or a built in invariant if callee was not in symbol table)
+        const auto &callee = to_code_type(function_call.function().type());
+        DATA_CHECK(callee.return_type().id() == ID_empty, "");
+      }
+    }
+  }
+}
+
+void validate_goto_modelt::check_called_functions()
+{
+  class test_for_function_addresst : public const_expr_visitort
+  {
+  private:
+    const irep_idt match_id{ID_address_of};
+
+  public:
+    std::set<irep_idt> identifiers;
+
+    void operator()(const exprt &expr) override
+    {
+      if(expr.id() == match_id)
+      {
+        const exprt &pointee = to_address_of_expr(expr).object();
+        if(pointee.id() == ID_symbol)
+          identifiers.insert(to_symbol_expr(pointee).get_identifier());
+      }
+    }
+
+    void clear()
+    {
+      identifiers.clear();
+    }
+  } test_for_function_address;
+
+  for(const auto &fun : function_map)
+  {
+    for(auto &instr : fun.second.body.instructions)
+    {
+      if(instr.is_function_call())
+      {
+        const auto &function_call = to_code_function_call(instr.code);
+        const irep_idt &identifier =
+          to_symbol_expr(function_call.function()).get_identifier();
+
+        DATA_CHECK(function_map.find(identifier) != function_map.end(),
+                   "every function call callee must be in the function map");
+      }
+
+      // to consider: excluding some of the 18 instruction types
+      // from this brute force check
+      const exprt &src{instr.code};
+      src.visit(test_for_function_address);
+      if(!test_for_function_address.identifiers.empty())
+      {
+        for(auto &identifier : test_for_function_address.identifiers)
+          DATA_CHECK(function_map.find(identifier) != function_map.end(),
```

Comment:
Does this work for pre-linking goto-programs?

Reply:
@martin-cs 
That got us thinking... and I rewrote the following check (which now has the comments below)
  /// Check that goto-programs that have a body end with an END_FUNCTION
  /// instruction.
  ///
  /// NB functions that are only declared, and not used will not appear in the
  /// function map. Functions that are declared only and used to e.g.
  /// initialise a function pointer will have no body.
void check_last_instruction();